ompi/communicator/comm_cid.c

```diff
@@ -315,7 +315,7 @@ static int ompi_comm_ext_cid_new_block (ompi_communicator_t *newcomm, ompi_commu
     opal_process_name_t *name_array = NULL;
     char *tag = NULL;
     size_t proc_count;
-    size_t cid_base;
+    size_t cid_base = 0;
```

Comment:
This fixes the warning but will break the code. Fact: we cannot have two CID blocks with the same base, or we will create conflicting set of CIDs (and each communicator is expected to have a unique CID). Thus, if the cid_base is not correctly extracted from the PMIX group we shall not continue, resulting in creating a second block with the same default (aka 0 with this patch) CID.

Reply:
@bosilca I belive you have misread the proposed fix. The act of just initializing a previously uninitialized variable cannot break the code by itself unless the initial value wrong. If initializing `cid_base = 0` is not the right initial value, then which value should the OP use?

Reply:
Maybe initialize  `cid_base = -1;`, and right after that value is used for read add `assert (cid_base >= 0);` ?

Reply:
I do believe it can, deterministically lead to faults. While the current code might work, just not deterministically. I explained the reasons above, if we create a new block with a random base (which will happen most of the time is we do explicitly initialize the variable), the code might seems to be working just fine. 

I'm not advocating on keeping the current code, I'm just pointing out that this fix might have a worst outcome than the original.

Reply:
To my understanding, an uninitialized variable does not really get a true random value. So as George said, are we lucky that the code works as-is?

@jiaxiyan Did you test this PR? Any failures?

Reply:
Reading the code, `cid_base` should be assigned on L383
```
    for (size_t i=0; i<nresults; i++) {
        if (PMIX_CHECK_KEY(&results[i], PMIX_GROUP_CONTEXT_ID)) {
            PMIX_VALUE_GET_NUMBER(rc, &results[i].value, cid_base, size_t);
            if(PMIX_SUCCESS != rc) {
                ret = opal_pmix_convert_status(rc);
                goto fn_exit;
            }
            break;
        }
    }
```

I wonder if it's possible that we don't hit this path and end up with the default value - would that be erroneous anyways?

Reply:
@bosilca If `PMIX_VALUE_GET_NUMBER()` fails to set `cid_base`, won't we go to `fn_exit` and therefore ignore the `cid_base` value?  I don't think it's possible for `nresults` to be less than 1, right?

If that's the case, then @dalcinl's suggestion of initially assigning the value to `-1` is probably good -- it makes it obvious that the value is invalid, but we'll only end up with `cid_base==-1` if `PMIX_VALUE_GET_NUMBER()` fails and we end up going to `fn_exit`.

Reply:
Updated initialization to -1 as suggested

Reply:
I'm not sure about Open MPI but some people prefer `SIZE_MAX` over `-1` for `size_t`.

Also this solution is not complete - you also need to check `cid_base` value and error out if it stays unassigned.

Reply:
@jiaxiyan Could you please add a line with

```c
assert (cid_base >= 0);
```
right BEFORE the line with
```c
 ompi_comm_extended_cid_block_initialize (new_block, cid_base, 0, 0);
```
That way, if `cid_base` is used but never set to anything meaningful, at least we know where to point our fingers.

Reply:
@dalcinl Note that `cid_base` is unsigned - I'm not sure `assert (cid_base >= 0);` will do anything.

The other option is to use a pointer `size_t *cid_base_ptr = NULL;` and only initialize it if `cid_base` is set correctly. Then we can do a NULL check on `cid_base_ptr`.

Reply:
Sorry, it is `size_t`!, then yes, it would be better to 

```c
size_t cid_base  = SIZE_MAX;
```
and then
```c
assert(SIZE_MAX != cid_base)
```
Would that work?

Reply:
If there's no meaningful sentinel value (e.g., because it's unsigned), you could just add another variable, such as `bool cid_base_set = false`, and only set it to `true` when the `PMIX_VALUE_GET_NUMBER()` returns successfully.  Then you can check `cid_base_set` before calling `ompi_comm_extend_cid_block_initialize()`.

It's not the most elegant thing in the world, but it is effective, easy to read/grok, and 100% safe.

Reply:
> Then you can check `cid_base_set`

If not, would it make sense to do cleanup and error with let say `ERR_INTERN` or `OMPI_ERR_OUT_OF_RESOURCE` ? Better than a naive `assert()`, right?

Reply:
Yes, we should not do `assert()`.  Realistically, if `PMIX_VALUE_GET_NUMBER()` fails, the current code *already* handles that case:

```c
    for (size_t i=0; i<nresults; i++) {
        if (PMIX_CHECK_KEY(&results[i], PMIX_GROUP_CONTEXT_ID)) {
            PMIX_VALUE_GET_NUMBER(rc, &results[i].value, cid_base, size_t);
            if(PMIX_SUCCESS != rc) {
                ret = opal_pmix_convert_status(rc);
                goto fn_exit;
            }
            break;
        }
    }
```

I'm guessing the compiler is just not smart enough / does not have enough context to know that we should *always* execute `PMIX_VALUE_GET_NUMBER()` here.

That being said, adding a belt-and-suspenders check for ensure that `cid_base` is *actually* set -- although likely redundant -- isn't harmful, and adds a bit of defensiveness in the code, which is not a bad thing.

Reply:
You really don't want to use an `assert` here as that only applies during a debug build, and this problem is just as likely to hit in non-debug scenarios. The only real way you would encounter it is in an environment that does not support `PMIx_Group` operations - which means a direct launch scenario like Slurm. In those cases, you'll see an error returned from the group op, and you'll know that the op failed and cid_base wasn't set by just looking at the return code.

In other words, all this is pretty much wasted breath, guys. You check the return from `PMIx_Group_construct`, and if it is an error, then you go to the error return. If not, then you check the error code from `PMIX_VALUE_GET_NUMBER` - and if that is success, then the cid_base is good.

You still need to init the variable to shut up the compiler, but all the angst is unnecessary.

Reply:
`PMIX_VALUE_GET_NUMBER` is a macro:
```c
#define PMIX_VALUE_GET_NUMBER(s, m, n, t)               \
    do {                                                \
        (s) = PMIX_SUCCESS;                             \
        if (PMIX_SIZE == (m)->type) {                   \
            (n) = (t)((m)->data.size);                  \
        } else if (PMIX_INT == (m)->type) {             \
            (n) = (t)((m)->data.integer);               \
        } else if (PMIX_INT8 == (m)->type) {            \
            (n) = (t)((m)->data.int8);                  \
        } else if (PMIX_INT16 == (m)->type) {           \
            (n) = (t)((m)->data.int16);                 \
        } else if (PMIX_INT32 == (m)->type) {           \
            (n) = (t)((m)->data.int32);                 \
        } else if (PMIX_INT64 == (m)->type) {           \
            (n) = (t)((m)->data.int64);                 \
        } else if (PMIX_UINT == (m)->type) {            \
            (n) = (t)((m)->data.uint);                  \
        } else if (PMIX_UINT8 == (m)->type) {           \
            (n) = (t)((m)->data.uint8);                 \
        } else if (PMIX_UINT16 == (m)->type) {          \
            (n) = (t)((m)->data.uint16);                \
        } else if (PMIX_UINT32 == (m)->type) {          \
            (n) = (t)((m)->data.uint32);                \
        } else if (PMIX_UINT64 == (m)->type) {          \
            (n) = (t)((m)->data.uint64);                \
        } else if (PMIX_FLOAT == (m)->type) {           \
            (n) = (t)((m)->data.fval);                  \
        } else if (PMIX_DOUBLE == (m)->type) {          \
            (n) = (t)((m)->data.dval);                  \
        } else if (PMIX_PID == (m)->type) {             \
            (n) = (t)((m)->data.pid);                   \
        } else if (PMIX_PROC_RANK == (m)->type) {       \
            (n) = (t)((m)->data.rank);                  \
        } else if (PMIX_STATUS == (m)->type) {          \
            (n) = (t)((m)->data.status);                \
        } else {                                        \
            (s) = PMIX_ERR_BAD_PARAM;                   \
        }                                               \
    } while(0)
```
As you can see, it returns an error if the number cannot be provided. It does _not_ set the number to a value in that case because we have no idea what value would constitute an error. So only the error code is relevant.

Compiler is just flagging that the value isn't set in an error case. Question is: how is `cid_base` being used in that situation...and why?

Reply:
@rhc all error case are correctly handled. And no the compiler does not complain about the value not being set in the error cases, because error cases lead to `fn_exit` with no use of `cid_base`.

The compiler complains about the case where `nresults` is zero or no returned info matches PMIX_GROUP_CONTEXT_ID, so cid_base remains uninitialized when the code reaches `ompi_comm_extended_cid_block_initialize`.

In that case setting cid_base to a predefined value will lead to more or less badness: from the worst, aka. colliding blocks of cid, to the best, aka. a steep performance degradation increasing as we create more communicators due to the need to find a cid available everywhere (cid that starts with a default predefined value, so we beat the same cid range all the time).

The current code, with the non initialized cid_base, has a better chance of success, because at least one of the processes might bring a different value. The solution to go for a default value could work for as long as we inform the user that we are defaulting to a predefined value and that might have significant performance implications on communicator creation.

Reply:
Hmmm...I see. Of course, those conditions (nresults=0 or no returned context ID, but group construct is successful) cannot happen, but that isn't something the compiler can know.

It's hard to find a value you can set to indicate invalid. Per request of @hppritcha, PRRTE starts at `UINT32_MAX` and counts downward. IIRC, somewhere OMPI starts at 0 and counts upward?

Sounds like @jsquyres suggestion is probably the best - use a boolean to indicate that a value was found. If not found, then perhaps the best solution is to have each proc generate its own random number? Would solve the "same range" issue, though I have no idea of what the overall impact might be.